keps/sig-scheduling/20200114-multi-scheduling-profiles.md

```diff
@@ -244,19 +246,32 @@ the scheduler documentation.
 
 ### Test Plan
 
-TODO
+The following tests need to be in place:
 
-### Graduation Criteria
+- **Unit Tests**: For Component Config API conversion and defaults, core scheduler and scheduler
+instantiation. Current tests that use a default scheduler (or default framework) should continue
+passing with minimal changes.
+- **Integration tests**: Current tests with a default scheduler should continue passing with minimal
+changes. We need new tests exercising more than one profile, in which:
+    - Each profile would favor specific pods, so that we can verify assignment.
+    - Pods get binding events for the selected scheduler name.
+    - Pods that don't specify a scheduler name continue to be scheduled by the default profile.
```

Comment:
That is fair. I'm usually hesitant to add more e2e tests as they are slow. We can just replicate one case of the "integration" tests. WDYT?

Reply:
I added one

Reply:
But technically, it's difficult to verify an alpha feature using e2e, as we can't set the feature gate in an e2e cluster, am I right? @liggitt

Reply:
In this case, this is a feature that is both alpha-level and requires specific cluster configuration. e2e tests for alpha-level features typically add a `[Feature:$FEATURE_NAME]` tag to exclude themselves by default, then opt into testing that tag in jobs that enable the feature and the relevant configuration.

Given you cannot assume this specialized multi-profile test configuration exists in arbitrary test setups, it would be good to describe the e2e job that will be used to test this, who will own it, etc

Reply:
I doesn't look like we have any job that exercises `kubescheduler.config.k8s.io/v1alpha1`. Does that mean that we need a whole new job? If so, it's fair to say that it should be owned by sig-scheduling.

Reply:
> I doesn't look like we have any job that exercises `kubescheduler.config.k8s.io/v1alpha1`. Does that mean that we need a whole new job? If so, it's fair to say that it should be owned by sig-scheduling.

Probably? Would suggest asking for recommendations for how to test this in sig-testing

Reply:
I think having a whole new job for one test of an alpha feature is almost an overkill. I'm moving this to Beta. Please let me know if you feel strongly about the E2E test.

Reply:
It sounds like the difficulty is not the alpha nature of the feature, but that a specific configuration is required. Based on the design of the feature, it sounds like that will be required whether the feature is alpha or beta. I would expect an e2e test before shipping this at any level.

Reply:
Why wouldn't integration tests be trusted in this scenario?

We basically need to guarantee 2 things:

- We are parsing configuration correctly
- We are scheduling based on the scheduler name.

Those 2 things can be guaranteed with integration tests (we only need to interact with an apiserver). I think E2E tests would be required if there was another component that needs to react. But there is no other component involved in this feature.

Reply:
> Why wouldn't integration tests be trusted in this scenario?

Are you proposing that we not have any e2e tests for this feature?

> We basically need to guarantee 2 things:
> 
> * We are parsing configuration correctly
> * We are scheduling based on the scheduler name.
>
> Those 2 things can be guaranteed with unit and integration tests (we only need to interact with an apiserver). I think E2E tests would be required if there was another component that needs to react. But there is no other component involved in this feature.

How would we test that the multiple schedulers correctly share the pool of knowledge about nodes' resources (called out as a benefit in this design)

That means you need to mock kubelet updates to node resource status as well, since we cannot run kubelets in integration tests. That starts to become a fair amount of mock behavior embedded in the test.

Reply:
There seems to be a confusion here.
We propose that there is only one kube-scheduler binary, running multiple profiles.
Node resources have always being calculated from the pod information as provided by the apiserver.

Reply:
Clarifications added.